### PR TITLE
add list secrets functionality, which returns keys given a path

### PR DESCRIPTION
**Pull request by marctreyes**
Uses GET version with query param list=true vs using LIST http method.

https://www.vaultproject.io/api-docs#api-operations
https://www.vaultproject.io/api/secret/kv/kv-v1#list-secrets

related https://github.com/Banno/vault4s/pull/207

### CODE REVIEW

**Review by gutiory — APPROVED** [review]

**Review by alejandrohdezma — APPROVED** [review]

File: core/src/main/scala/com/banno/vault/Vault.scala [hunk 1]
```diff
@@ -86,6 +86,27 @@ object Vault {
...
+      )
+    for {
+      result <- 
+        F.adaptError(client.expect[VaultKeys](request)(jsonOf[F, VaultKeys])) {
+          case InvalidMessageBodyFailure(_, Some(cause: DecodingFailure)) =>
+            InvalidMessageBodyFailure("Could not decode vault list secrets response", cause.some)
+        }.handleErrorWith { e =>
+          F.raiseError(VaultRequestError(request = request, cause = e.some, extra = s"tokenLength=${token.length}".some))
+        }
+    } yield result
```
Comment (diesalbla) [inline core/src/main/scala/com/banno/vault/Vault.scala:107]: No need of for-comprehension if there is a single item. 
```suggestion
      F.adaptError(client.expect[VaultKeys](request)(jsonOf[F, VaultKeys])) {
        case InvalidMessageBodyFailure(_, Some(cause: DecodingFailure)) =>
          InvalidMessageBodyFailure("Could not decode vault list secrets response", cause.some)
      }.handleErrorWith { e =>
        F.raiseError(VaultRequestError(request = request, cause = e.some, extra = s"tokenLength=${token.length}".some))
      }
```

File: core/src/main/scala/com/banno/vault/models/VaultKeys.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  implicit val vaultKeysDecoder: Decoder[VaultKeys] =
+    Decoder.instance[VaultKeys] { c =>
+      Decoder.resultInstance.map(
+        c.downField("data").get[List[String]]("keys")
+      )(VaultKeys.apply)
+    }
+
+  implicit val VaultKeysEq : Eq[VaultKeys] = Eq.instance[VaultKeys]((vt1, vt2) =>
+    vt1.keys === vt2.keys
+  )
```
Comment (diesalbla) [inline core/src/main/scala/com/banno/vault/models/VaultKeys.scala:36]:
```suggestion
  implicit val VaultKeysEq : Eq[VaultKeys] = Eq.fromUniversalEquals[VaultKeys]
```

**Review by diesalbla — APPROVED** [review]

File: core/src/main/scala/com/banno/vault/Vault.scala [hunk 3]
```diff
@@ -86,6 +86,24 @@ object Vault {
...
+        method = Method.GET,
+        uri = vaultUri.withPath(s"/v1/$newSecretPath").withQueryParam("list", "true"),
+        headers = Headers.of(Header("X-Vault-Token", token))
+      )
+    F.adaptError(client.expect[VaultKeys](request)(jsonOf[F, VaultKeys])) {
+      case InvalidMessageBodyFailure(_, Some(cause: DecodingFailure)) =>
+        InvalidMessageBodyFailure("Could not decode vault list secrets response", cause.some)
+    }.handleErrorWith { e =>
+      F.raiseError(VaultRequestError(request = request, cause = e.some, extra = s"tokenLength=${token.length}".some))
+    }
```
Comment (diesalbla) [inline core/src/main/scala/com/banno/vault/Vault.scala:104]:
```suggestion
    }.adaptError { e =>
      VaultRequestError(request = request, cause = e.some, extra = s"tokenLength=${token.length}".some)
    }
```

Comment (marctreyes) [reply core/src/main/scala/com/banno/vault/Vault.scala:104]: I'll consolidate this with the `F.adaptError` body above.

**Review by trentio — APPROVED** [review]

**Pull request merged**